src/charm.py

```diff
@@ -19,6 +24,169 @@ class KratosCharm(CharmBase):
+        if not self._container.get_plan().to_dict():
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self._container.push(self._config_file_path, self._config, make_dirs=True)
+            with open("src/identity.default.schema.json", encoding="utf-8") as schema_file:
+                schema = schema_file.read()
+                self._container.push(self._identity_schema_file_path, schema, make_dirs=True)
+            self._container.add_layer(self._container_name, self._pebble_layer, combine=True)
+            logger.info("Pebble plan updated with new configuration, replanning")
+            self._container.replan()
+        else:
+            # Compare changes in kratos config
+            current_config = self._container.pull(self._config_file_path).read()
+            if current_config != self._config:
+                self.unit.status = MaintenanceStatus("Updating Kratos Config")
+                self._container.push(self._config_file_path, self._config, make_dirs=True)
+                logger.info("Updated kratos config")
+                self._container.restart()
+
+    def _get_database_relation_info(self) -> dict:
+        """Get database info from relation data bag."""
+        relation_id = self.database.relations[0].id
+        relation_data = self.database.fetch_relation_data()[relation_id]
+
+        return {
+            "username": relation_data["username"],
+            "password": relation_data["password"],
+            "endpoints": relation_data["endpoints"],
+        }
+
+    def _update_container(self, event) -> None:
+        """Update configs, pebble layer and run database migration."""
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Kratos container. Deferring event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Kratos container")
+            return
+
+        if not self.model.relations["pg-database"]:
+            event.defer()
+            logger.error("Missing required relation with postgresql")
+            self.model.unit.status = BlockedStatus("Missing required relation with postgresql")
+            return
+
+        if not self.database.is_database_created():
+            event.defer()
+            logger.info("Missing database details. Deferring event.")
+            self.unit.status = WaitingStatus("Waiting for database creation")
+            return
+
+        try:
+            self._update_layer()
+        except ChangeError as err:
+            logger.error(str(err))
+            self.unit.status = BlockedStatus("Failed to replan")
+            return
+
+        if not self.unit.is_leader():
+            event.defer()
+            logger.info("Does not have leadership. Deferring event.")
+            self.unit.status = WaitingStatus("Waiting for leadership")
+            return
+
+        self._run_sql_migration()
+
+        self.unit.status = ActiveStatus()
```

Comment:
It feels like too many stuff is happening in this (+`_update_layer`) method. It is handling pebble readiness, changes in configuration and changes in the database. I think that it's better to break this into smaller functions, as we don't need to be doing all of those things on every event. E.g.:
- On pebble ready: update/push the config, push the layer, update status. This will probably raise errors as there will be no db, but we will be in a `BlockedStatus` so it's expected. We could defer the pebble ready event until the database is there, but IMHO that would complicate the flow without making the UX much better.
- On config changed: update/push the config, update status
- On database created: run migration, update/push the config, update status
- On database endpoint changed: update/push the config, update status

What do you think?

Reply:
The problem is that the order of events is not guaranteed. On database created could happen before on pebble ready. I am not observing config changed right now, but it could also happen before on pebble ready. I don't know for sure the state of the charm when an event is emitted. That's why I am coding in a very defensive manner. 
Regarding migration, this is an interim solution. I will open a separate PR for better handling migration.

Reply:
Still, if on every event you are forcing idempotent changes to the state of the container, you don't have to worry about the order of the events. For example if database_created comes before pebble_ready shouldn't matter as:
- The database_created hook will try migrating the db (it will succeed), updating the config (this will return without doing any work as the container will not be ready), restarting the service (this will return without doing any work as the container will not be ready), updating the status (this will set the status to maintenance or something).
- The pebble_ready hook will push the config (which will include the database config), push the pebble layer, update the status (this will set the status to active as both the container and the db will be ready).